samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java

```diff
@@ -77,20 +93,81 @@
    */
   public LocalApplicationRunner(SamzaApplication app, Config config) {
     this.appDesc = ApplicationDescriptorUtil.getAppDescriptor(app, config);
-    this.planner = new LocalJobPlanner(appDesc);
+    isAppModeBatch = new ApplicationConfig(config).getAppMode() == ApplicationConfig.ApplicationMode.BATCH;
+    coordinationUtils = getCoordinationUtils(config);
   }
 
   /**
    * Constructor only used in unit test to allow injection of {@link LocalJobPlanner}
    */
   @VisibleForTesting
-  LocalApplicationRunner(ApplicationDescriptorImpl<? extends ApplicationDescriptor> appDesc, LocalJobPlanner planner) {
+  LocalApplicationRunner(ApplicationDescriptorImpl<? extends ApplicationDescriptor> appDesc, Optional<CoordinationUtils> coordinationUtils) {
     this.appDesc = appDesc;
-    this.planner = planner;
+    isAppModeBatch = new ApplicationConfig(appDesc.getConfig()).getAppMode() == ApplicationConfig.ApplicationMode.BATCH;
+    this.coordinationUtils = coordinationUtils;
+  }
+
+  private Optional<CoordinationUtils> getCoordinationUtils(Config config) {
+    if (!isAppModeBatch) {
+      return Optional.empty();
+    }
+    JobCoordinatorConfig jcConfig = new JobCoordinatorConfig(config);
+    CoordinationUtils coordinationUtils = jcConfig.getCoordinationUtilsFactory().getCoordinationUtils(CoordinationConstants.APPLICATION_RUNNER_PATH_SUFFIX, PROCESSOR_ID, config);
+    return Optional.ofNullable(coordinationUtils);
+  }
+
+  /**
+   * @return LocalJobPlanner created
+   */
+  @VisibleForTesting
+  LocalJobPlanner getPlanner() {
+    boolean isAppModeBatch = new ApplicationConfig(appDesc.getConfig()).getAppMode() == ApplicationConfig.ApplicationMode.BATCH;
+    if (!isAppModeBatch) {
+      return new LocalJobPlanner(appDesc, PROCESSOR_ID);
+    }
+    CoordinationUtils coordinationUtils = null;
+    String runId = null;
+    if (this.coordinationUtils.isPresent()) {
+      coordinationUtils = this.coordinationUtils.get();
+    }
+    if (this.runId.isPresent()) {
+      runId = this.runId.get();
+    }
+    return new LocalJobPlanner(appDesc, coordinationUtils, PROCESSOR_ID, runId);
+  }
+
+
+  private void initializeRunId() {
+    try {
+      MetadataStore metadataStore = getMetadataStore();
+      if (coordinationUtils.isPresent() && metadataStore != null) {
+        runIdGenerator = Optional.of(new RunIdGenerator(coordinationUtils.get(), metadataStore));
+      }
+      if (!coordinationUtils.isPresent() || !isAppModeBatch || !runIdGenerator.isPresent()) {
+        LOG.warn("coordination utils or run id generator could not be created successfully!");
+        return;
+      }
+      runId = runIdGenerator.get().getRunId();
```

Comment:
runIdGenerator check can be removed and you can simplify this to `runId = runIdGenerator.ifPresent(RunIdGenerator::getRunId);`